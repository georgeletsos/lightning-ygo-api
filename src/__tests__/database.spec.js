require('dotenv').config();

const mongoose = require('mongoose');
const { MongoMemoryServer } = require('mongodb-memory-server');
const mongoServer = new MongoMemoryServer();
const database = require('../database');
const Card = require('../models/Card');
const mocks = require('./mocks');

beforeAll(async () => {
  const uri = await mongoServer.getUri();
  await mongoose.connect(uri, {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });
});

afterAll(async () => {
  await mongoose.disconnect();
  await mongoServer.stop();
});

it('`fetchMissingCards()` - should find any missing cards between the API and the database)', async () => {
  expect.assertions(1);

  const missingCards = await database.fetchMissingCards();
  expect(missingCards).not.toHaveLength(0);
}, 30000);

it('`convertMissingCards(missingCards)` - should convert any given missing cards based on the database schema', async () => {
  expect.assertions(1);

  const convertedMissingCards = await database.convertMissingCards(
    mocks.mockMissingCards
  );

  const convertedMissingCardsWithout_id = JSON.parse(
    JSON.stringify(convertedMissingCards)
  );
  convertedMissingCardsWithout_id.forEach(convertedMissingCard => {
    delete convertedMissingCard._id;
  });

  expect(convertedMissingCardsWithout_id).toMatchSnapshot();
}, 30000);

it('should save a card to the database', async () => {
  expect.assertions(10);

  const convertedMissingCards = await database.convertMissingCards(
    mocks.mockMissingCards
  );
  const convertedCard = convertedMissingCards[2];

  const savedCard = await Card.create(convertedCard);
  expect(savedCard.cardType).toEqual(convertedCard.cardType);
  expect(savedCard.name).toEqual(convertedCard.name);
  expect(savedCard.attribute).toEqual(convertedCard.attribute);
  expect(savedCard.level).toEqual(convertedCard.level);
  expect(savedCard.monsterType).toEqual(convertedCard.monsterType);
  expect(savedCard.types).toEqual(convertedCard.types);
  expect(savedCard.text).toEqual(convertedCard.text);
  expect(savedCard.atk).toEqual(convertedCard.atk);
  expect(savedCard.def).toEqual(convertedCard.def);
  expect(savedCard.image).toEqual(convertedCard.image);
}, 30000);
