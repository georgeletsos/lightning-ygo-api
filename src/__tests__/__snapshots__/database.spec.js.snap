// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`convertMissingCards(missingCards)\` - should convert any given missing cards based on the database schema 1`] = `
Array [
  Object {
    "atk": 500,
    "attribute": "light",
    "cardType": "monster",
    "def": 200,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584275991/lightning_ygo_api/card_images/art/652362.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584275989/lightning_ygo_api/card_images/big/652362.jpg",
      "id": "652362",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584275990/lightning_ygo_api/card_images/small/652362.jpg",
    },
    "level": 1,
    "monsterType": "reptile",
    "name": "Alien Ammonite",
    "text": "When this card is Normal Summoned, you can Special Summon 1 Level 4 or lower \\"Alien\\" monster from your Graveyard. Destroy it during the End Phase.",
    "types": Array [
      "tuner",
      "effect",
    ],
  },
  Object {
    "atk": 300,
    "attribute": "light",
    "cardType": "monster",
    "def": 800,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584275999/lightning_ygo_api/card_images/art/62437709.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584275997/lightning_ygo_api/card_images/big/62437709.jpg",
      "id": "62437709",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584275998/lightning_ygo_api/card_images/small/62437709.jpg",
    },
    "level": 2,
    "monsterType": "reptile",
    "name": "Alien Grey",
    "text": "FLIP: Place 1 A-Counter on 1 face-up monster on your opponent's side of the field. (If a monster with an A-Counter battles an \\"Alien\\" monster, it loses 300 ATK and DEF for each A-Counter during damage calculation only.) After this card has been flipped face-up, when it is destroyed by battle and sent to the Graveyard, draw 1 card.",
    "types": Array [
      "flip",
      "effect",
    ],
  },
  Object {
    "atk": 2400,
    "attribute": "dark",
    "cardType": "monster",
    "def": 1600,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276108/lightning_ygo_api/card_images/art/19204398.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276106/lightning_ygo_api/card_images/big/19204398.jpg",
      "id": "19204398",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276107/lightning_ygo_api/card_images/small/19204398.jpg",
    },
    "level": 8,
    "monsterType": "machine",
    "name": "Ally of Justice Light Gazer",
    "text": "1 Tuner + 1 or more non-Tuner monsters
This card gains 200 ATK for each LIGHT monster in your opponent's Graveyard.",
    "types": Array [
      "synchro",
      "effect",
    ],
  },
  Object {
    "atk": 1900,
    "attribute": "earth",
    "cardType": "monster",
    "def": 1200,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276182/lightning_ygo_api/card_images/art/32181268.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276180/lightning_ygo_api/card_images/big/32181268.jpg",
      "id": "32181268",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276181/lightning_ygo_api/card_images/small/32181268.jpg",
    },
    "level": 4,
    "monsterType": "rock",
    "name": "Amano-Iwato",
    "text": "Cannot be Special Summoned. Monsters cannot activate their effects, except Spirit monsters. Once per turn, during the End Phase, if this card was Normal Summoned or flipped face-up this turn: Return this card to the hand.",
    "types": Array [
      "spirit",
      "effect",
    ],
  },
  Object {
    "atk": 2900,
    "attribute": "dark",
    "cardType": "monster",
    "def": 2500,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276303/lightning_ygo_api/card_images/art/75380687.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276300/lightning_ygo_api/card_images/big/75380687.jpg",
      "id": "75380687",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276301/lightning_ygo_api/card_images/small/75380687.jpg",
    },
    "level": 8,
    "monsterType": "dragon",
    "name": "Amulet Dragon",
    "text": "\\"Dark Magician\\" + 1 Dragon monster
Must be Fusion Summoned with the above Fusion Materials or with \\"The Eye of Timaeus\\". If this card is Special Summoned: Target any number of Spells in any GY(s) (min.1); banish those targets, and if you do, this card gains 100 ATK for each card banished by this effect. If this card is destroyed: You can target 1 Spellcaster monster in your GY; Special Summon that target.",
    "types": Array [
      "fusion",
      "effect",
    ],
  },
  Object {
    "atk": 1800,
    "attribute": "earth",
    "cardType": "monster",
    "def": 500,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276408/lightning_ygo_api/card_images/art/39303359.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276405/lightning_ygo_api/card_images/big/39303359.jpg",
      "id": "39303359",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276407/lightning_ygo_api/card_images/small/39303359.jpg",
    },
    "level": 4,
    "monsterType": "machine",
    "name": "Ancient Gear Knight",
    "text": "This card is treated as a Normal Monster while face-up on the field or in the Graveyard. While this card is a Normal Monster on the field, you can Normal Summon it to have it become an Effect Monster with this effect.
‚óè If this card attacks, your opponent cannot activate any Spell/Trap Cards until the end of the Damage Step.",
    "types": Array [
      "gemini",
      "effect",
    ],
  },
  Object {
    "atk": 1900,
    "attribute": "earth",
    "cardType": "monster",
    "def": 1600,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276478/lightning_ygo_api/card_images/art/87979586.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276475/lightning_ygo_api/card_images/big/87979586.jpg",
      "id": "87979586",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276477/lightning_ygo_api/card_images/small/87979586.jpg",
    },
    "level": 4,
    "monsterType": "plant",
    "name": "Angel Trumpeter",
    "text": "This beautiful, fairy-like flower generates a delusional mist to keep intruders away from the deepest parts of the forest, where sacred beasts dwell.",
    "types": Array [
      "tuner",
      "normal",
    ],
  },
  Object {
    "atk": 800,
    "attribute": "earth",
    "cardType": "monster",
    "def": 800,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276770/lightning_ygo_api/card_images/art/87798440.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276767/lightning_ygo_api/card_images/big/87798440.jpg",
      "id": "87798440",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584276769/lightning_ygo_api/card_images/small/87798440.jpg",
    },
    "level": 3,
    "monsterType": "warrior",
    "name": "Armor Breaker",
    "text": "Once per turn, you can equip this card to a Warrior-Type monster you control, OR unequip it to Special Summon this card in face-up Attack Position. While equipped to a monster by this effect, when that monster inflicts Battle Damage to your opponent, destroy 1 card on the field. (A monster can only be equipped with 1 Union Monster at a time. If the equipped monster would be destroyed by battle, this card is destroyed instead.)",
    "types": Array [
      "union",
      "effect",
    ],
  },
  Object {
    "atk": 1380,
    "attribute": "dark",
    "cardType": "monster",
    "def": 1530,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1586941868/lightning_ygo_api/card_images/art/6840573.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1586941865/lightning_ygo_api/card_images/big/6840573.jpg",
      "id": "6840573",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1586941866/lightning_ygo_api/card_images/small/6840573.jpg",
    },
    "level": 5,
    "monsterType": "fiend",
    "name": "Barox",
    "text": "\\"Frenzied Panda\\" + \\"Ryu-Kishin\\"",
    "types": Array [
      "fusion",
    ],
  },
  Object {
    "atk": null,
    "attribute": null,
    "cardType": "spell",
    "def": null,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584277697/lightning_ygo_api/card_images/art/41426869.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584277692/lightning_ygo_api/card_images/big/41426869.jpg",
      "id": "41426869",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584277695/lightning_ygo_api/card_images/small/41426869.jpg",
    },
    "level": null,
    "monsterType": null,
    "name": "Black Illusion Ritual",
    "text": "This card is used to Ritual Summon \\"Relinquished\\". You must also Tribute a monster from your hand or field whose Level is 1 or more.",
    "types": Array [
      "ritual",
    ],
  },
  Object {
    "atk": 0,
    "attribute": "dark",
    "cardType": "monster",
    "def": 0,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1589816105/lightning_ygo_api/card_images/art/67284108.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1589816103/lightning_ygo_api/card_images/big/67284108.jpg",
      "id": "67284108",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1589816104/lightning_ygo_api/card_images/small/67284108.jpg",
    },
    "level": 1,
    "monsterType": "zombie",
    "name": "Black Sheep Token",
    "text": "Special Summoned by the effect of \\"Scapeghost\\".",
    "types": Array [
      "token",
    ],
  },
  Object {
    "atk": 3000,
    "attribute": "light",
    "cardType": "monster",
    "def": 2500,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584278257/lightning_ygo_api/card_images/art/53183600.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584278254/lightning_ygo_api/card_images/big/53183600.jpg",
      "id": "53183600",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584278255/lightning_ygo_api/card_images/small/53183600.jpg",
    },
    "level": 8,
    "monsterType": "dragon",
    "name": "Blue-Eyes Toon Dragon",
    "text": "Cannot be Normal Summoned/Set. Must first be Special Summoned (from your hand) by Tributing 2 monsters, while you control \\"Toon World\\". Cannot attack the turn it is Special Summoned. You must pay 500 LP to declare an attack with this monster. If \\"Toon World\\" on the field is destroyed, destroy this card. Can attack your opponent directly, unless they control a Toon monster, in which case this card must target a Toon monster for its attacks.",
    "types": Array [
      "toon",
      "effect",
    ],
  },
  Object {
    "atk": "?",
    "attribute": "dark",
    "cardType": "monster",
    "def": "?",
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584279252/lightning_ygo_api/card_images/art/64599569.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584279250/lightning_ygo_api/card_images/big/64599569.jpg",
      "id": "64599569",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584279251/lightning_ygo_api/card_images/small/64599569.jpg",
    },
    "level": 9,
    "monsterType": "machine",
    "name": "Chimeratech Overdragon",
    "text": "\\"Cyber Dragon\\" + 1+ Machine monsters
Must be Fusion Summoned. If this card is Fusion Summoned: Send all other cards you control to the GY. The original ATK/DEF of this card each become equal to the number of Fusion Materials used for its Fusion Summon x 800. Each turn, this card can attack your opponent's monsters a number of times equal to the number of Fusion Materials used for its Fusion Summon.",
    "types": Array [
      "fusion",
      "effect",
    ],
  },
  Object {
    "atk": 2700,
    "attribute": "earth",
    "cardType": "monster",
    "def": 1800,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1584292110/lightning_ygo_api/card_images/art/45815891.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1584292106/lightning_ygo_api/card_images/big/45815891.jpg",
      "id": "45815891",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1584292107/lightning_ygo_api/card_images/small/45815891.jpg",
    },
    "level": 7,
    "monsterType": "fiend",
    "name": "Scrap Archfiend",
    "text": "1 Tuner + 1 or more non-Tuner monsters",
    "types": Array [
      "synchro",
    ],
  },
  Object {
    "atk": null,
    "attribute": null,
    "cardType": "trap",
    "def": null,
    "image": Object {
      "art": "https://res.cloudinary.com/georgeletsos/image/upload/v1588871867/lightning_ygo_api/card_images/card_back_warning.jpg",
      "big": "https://res.cloudinary.com/georgeletsos/image/upload/v1588871867/lightning_ygo_api/card_images/card_back_warning.jpg",
      "id": "0",
      "small": "https://res.cloudinary.com/georgeletsos/image/upload/v1588871867/lightning_ygo_api/card_images/card_back_warning.jpg",
    },
    "level": null,
    "monsterType": null,
    "name": "Passion of Baimasse",
    "text": "When your opponent activates a monster effect, while you control an \\"Earthbound Immortal\\" monster: Negate the activation and destroy it.",
    "types": Array [
      "counter",
    ],
  },
]
`;
